src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,413 @@
+        /// </summary>
+        /// <remarks>
+        /// Scanning ports and testing for devices may affect unrelated devices. It is advisable to exclude ports known to contain other hardware from this scan.
+        /// A board won't be found if its port is already open (by the same or a different process).
+        /// </remarks>
+        /// <param name="comPorts">List of com ports. Can be used with <see cref="SerialPort.GetPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        public static ArduinoBoard? FindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates)
+        {
+            foreach (var port in comPorts)
+            {
+                foreach (var baud in baudRates)
+                {
+                    ArduinoBoard? b = null;
+                    try
+                    {
+                        b = new ArduinoBoard(port, baud);
+                        b.Initialize();
+                        return b;
+                    }
+                    catch (Exception x) when (x is NotSupportedException || x is TimeoutException || x is IOException || x is UnauthorizedAccessException)
+                    {
+                        b?.Dispose();
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Searches all available com ports for an Arduino device.
+        /// </summary>
+        /// <returns>A board, already open and initialized. Null if none was found.</returns>
+        /// <remarks>
+        /// Scanning serial ports may affect unrelated devices. If there are problems, use the
+        /// <see cref="FindBoard(System.Collections.Generic.IEnumerable{string},System.Collections.Generic.IEnumerable{int})"/> overload excluding ports that shall not be tested.
+        /// </remarks>
+        public static ArduinoBoard? FindBoard()
+        {
+            return FindBoard(SerialPort.GetPortNames(), CommonBaudRates());
+        }
+
+        /// <summary>
+        /// Returns a list of commonly used baud rates.
+        /// </summary>
+        public static List<int> CommonBaudRates()
+        {
+            return new List<int>()
+            {
+                9600,
+                19200,
+                18400,
+                57600,
+                115200,
+                230400,
+                250000,
+                500000,
+                1000000,
+                2000000,
+            };
+        }
+
+        /// <summary>
+        /// Initialize the board connection. This must be called before any other methods of this class.
+        /// </summary>
+        /// <exception cref="NotSupportedException">The Firmata firmware on the connected board is too old.</exception>
+        public virtual void Initialize()
```

Comment:
I still see this guy public, I thought we said we were going to add a _initialized field and check on every method that needs it and call Initialize if it hasn't been called yet. That way you a) can remove Initialize from the public surface area and b) you avoid a very common pitfall folks will have when they forget to call initialize.

Reply:
I forgot about this... But I'd rather have to keep it public and throw if _initialzed is false, to force correct usage. This should be explicly called, to make sure exceptions are handled in one single place. I think it would be bad if we get exceptions from initialization errors from all possible methods.

Reply:
All exceptions will really come from Initialize in the callstack though. This pattern of lazy initialization is used elsewhere in the repo, for example: https://github.com/dotnet/iot/blob/b5bb684bbf00ebd266018cfffde3dcbde26f0538/src/System.Device.Gpio/System/Device/Spi/Devices/UnixSpiDevice.cs#L44

I think it is not great to design an API that we know will require a method to be called first, but doesn't call it and instead expects the user to have called it. The only argument that could be made is that you get a minor perf hit since you now  have to check on every action if it has been initialized or not, but seems like you are going to do that anyway in order to throw the exception.

Reply:
The main reason for separating the ctor from init is actually testability. If you want to write unit tests for individual functions with Mocks, you do not want to write up the mocks for the initialization sequence in each test case. (There's a few other things missing here to do be able to do that, but that's up for another PR).

Reply:
We shouldn't be designing Api surface for testing purposes, at least not when that means affecting the usability of the types. Primary focus should be usability, for tests we could use other techniques like InternalsVisibleTo or making methods virtual to override them, or just using Moq framework. For this particular case, we could move the actual implementation that doesn't call Initialize to an internal method, then use InternalsVisibleTo for the tests and use Moq to change the implementation and call the internal method directly without calling Initialize.